File: sdks/go/worker.go
package blok

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/nats-io/nats.go"
	"github.com/nats-io/nats.go/jetstream"
)

// WorkerConfig holds the configuration for a NATS JetStream worker.
type WorkerConfig struct {
	// Servers is a list of NATS server URLs (default: ["localhost:4222"]).
	Servers []string

	// StreamName is the JetStream stream name (default: "blok-worker").
	StreamName string

	// Token is the NATS authentication token (optional).
	Token string

	// User is the NATS username (optional).
	User string

	// Pass is the NATS password (optional).
	Pass string

	// Concurrency is the maximum number of concurrent job handlers (default: 1).
	Concurrency int

	// MaxRetries is the maximum number of delivery attempts (default: 3).
	MaxRetries int

	// AckWait is the timeout for job processing before NATS redelivers (default: 30s).
	AckWait time.Duration

	// Queues is the list of queue names to subscribe to.
	Queues []string
}

// DefaultWorkerConfig returns a WorkerConfig with sensible defaults.
func DefaultWorkerConfig() WorkerConfig {
	return WorkerConfig{
		Servers:     []string{"localhost:4222"},
		StreamName:  "blok-worker",
		Concurrency: 1,
		MaxRetries:  3,
		AckWait:     30 * time.Second,
	}
}

// LoadWorkerConfigFromEnv loads worker configuration from environment variables.
//
// Environment variables:
//   - NATS_SERVERS: Comma-separated NATS server URLs (default: localhost:4222)
//   - NATS_TOKEN: Authentication token (optional)
//   - NATS_USER: Username (optional)
//   - NATS_PASS: Password (optional)
//   - NATS_STREAM_NAME: JetStream stream name (default: blok-worker)
//   - WORKER_CONCURRENCY: Max concurrent jobs (default: 1)
//   - WORKER_MAX_RETRIES: Max delivery attempts (default: 3)
//   - WORKER_ACK_WAIT: Job timeout duration (default: 30s)
//   - WORKER_QUEUES: Comma-separated queue names to subscribe to
func LoadWorkerConfigFromEnv() WorkerConfig {
	cfg := DefaultWorkerConfig()

	if v := os.Getenv("NATS_SERVERS"); v != "" {
		cfg.Servers = strings.Split(v, ",")
	}

	if v := os.Getenv("NATS_TOKEN"); v != "" {
		cfg.Token = v
	}

	if v := os.Getenv("NATS_USER"); v != "" {
		cfg.User = v
	}

	if v := os.Getenv("NATS_PASS"); v != "" {
		cfg.Pass = v
	}

	if v := os.Getenv("NATS_STREAM_NAME"); v != "" {
		cfg.StreamName = v
	}

	if v := os.Getenv("WORKER_CONCURRENCY"); v != "" {
		if n, err := strconv.Atoi(v); err == nil && n > 0 {
			cfg.Concurrency = n
		}
	}

	if v := os.Getenv("WORKER_MAX_RETRIES"); v != "" {
		if n, err := strconv.Atoi(v); err == nil && n >= 0 {
			cfg.MaxRetries = n
		}
	}

	if v := os.Getenv("WORKER_ACK_WAIT"); v != "" {
		if d, err := time.ParseDuration(v); err == nil && d > 0 {
			cfg.AckWait = d
		}
	}

	if v := os.Getenv("WORKER_QUEUES"); v != "" {
		cfg.Queues = strings.Split(v, ",")
		for i, q := range cfg.Queues {
			cfg.Queues[i] = strings.TrimSpace(q)
		}
	}

	return cfg
}

// JobMessage represents a job received from a NATS worker queue.
type JobMessage struct {
	// ID is the unique job identifier.
	ID string

	// Queue is the queue name this job came from.
	Queue string

	// Data is the raw JSON job payload.
	Data json.RawMessage

	// Headers are the message headers.
	Headers map[string]string

	// Attempt is the current delivery attempt (0-based).
	Attempt int

	// MaxRetries is the maximum number of retries configured.
	MaxRetries int
}

// DataAs unmarshals the job data into the target struct.
func (j *JobMessage) DataAs(target interface{}) error {
	return json.Unmarshal(j.Data, target)
}

// DataMap returns the job data as a map, or nil if not a valid JSON object.
func (j *JobMessage) DataMap() map[string]interface{} {
	var m map[string]interface{}
	if err := json.Unmarshal(j.Data, &m); err != nil {
		return nil
	}
	return m
}

// JobHandler is the function signature for handling jobs.
type JobHandler func(ctx context.Context, job *JobMessage) error

// Worker processes background jobs from NATS JetStream queues.
type Worker struct {
	config   WorkerConfig
	registry *NodeRegistry
	handlers map[string]JobHandler

	nc   *nats.Conn
	js   jetstream.JetStream
	mu   sync.Mutex
	subs []jetstream.ConsumeContext
}

// NewWorker creates a new NATS JetStream worker.
func NewWorker(registry *NodeRegistry, config WorkerConfig) *Worker {
	return &Worker{
		config:   config,
		registry: registry,
		handlers: make(map[string]JobHandler),
	}
}

// Handle registers a custom handler for a queue.
func (w *Worker) Handle(queue string, handler JobHandler) {
	w.handlers[queue] = handler
}

// HandleNode registers a handler that routes jobs to a registered node.
// The job data becomes the node's input context (ctx.request.body).
func (w *Worker) HandleNode(queue, nodeName string) {
	w.handlers[queue] = func(ctx context.Context, job *JobMessage) error {
		// Build an ExecutionRequest matching the SDK's /execute contract
		req := &ExecutionRequest{
			Node: NodeConfig{
				Name:   nodeName,
				Config: job.DataMap(),
			},
			Context: Context{
				ID: job.ID,
				Request: Request{
					Body:    job.DataMap(),
					Headers: job.Headers,
					Params: map[string]string{
						"queue":   job.Queue,
						"jobId":   job.ID,
						"attempt": strconv.Itoa(job.Attempt),
					},
					Query: map[string]string{},
				},
				Vars: map[string]interface{}{
					"_worker_job": map[string]interface{}{
						"id":         job.ID,
						"queue":      job.Queue,
						"attempts":   strconv.Itoa(job.Attempt),
						"maxRetries": strconv.Itoa(job.MaxRetries),
					},
				},
			},
		}

		result := w.registry.Execute(req)
		if !result.Success {
			errMsg := "node execution failed"
			if errs, ok := result.Errors.(map[string]string); ok {
				if msg, ok := errs["message"]; ok {
					errMsg = msg
				}
			}
			return fmt.Errorf("%s", errMsg)
		}

		return nil
	}
}

// Start connects to NATS, ensures streams/consumers, and starts processing jobs.
// It blocks until the context is cancelled.
func (w *Worker) Start(ctx context.Context) error {
	// Build NATS connection options
	opts := []nats.Option{
		nats.Name("blok-worker"),
	}

	if w.config.Token != "" {
		opts = append(opts, nats.Token(w.config.Token))
	}
	if w.config.User != "" && w.config.Pass != "" {
		opts = append(opts, nats.UserInfo(w.config.User, w.config.Pass))
	}

	// Connect to NATS
	nc, err := nats.Connect(strings.Join(w.config.Servers, ","), opts...)
	if err != nil {
		return fmt.Errorf("failed to connect to NATS: %w", err)
	}
	w.nc = nc

	// Create JetStream context
	js, err := jetstream.New(nc)
	if err != nil {
		return fmt.Errorf("failed to create JetStream context: %w", err)
	}
	w.js = js

	log.Printf("[Worker] Connected to NATS: %s", strings.Join(w.config.Servers, ", "))

	// Process each queue
	for queue, handler := range w.handlers {
		if err := w.startQueue(ctx, queue, handler); err != nil {
			return fmt.Errorf("failed to start queue %s: %w", queue, err)
		}
	}

	log.Printf("[Worker] Processing %d queue(s), concurrency=%d", len(w.handlers), w.config.Concurrency)

	// Block until context is done
	<-ctx.Done()

	log.Println("[Worker] Shutting down...")
	return w.Stop()
}

// Stop gracefully stops all consumers and disconnects from NATS.
func (w *Worker) Stop() error {
	w.mu.Lock()
	defer w.mu.Unlock()

	// Stop all consume contexts
	for _, sub := range w.subs {
		sub.Stop()
	}
	w.subs = nil

	// Drain NATS connection
	if w.nc != nil {
		if err := w.nc.Drain(); err != nil {
			log.Printf("[Worker] Error draining NATS connection: %v", err)
		}
	}

	log.Println("[Worker] Stopped")
	return nil
}

// startQueue ensures stream/consumer exist and starts consuming for one queue.
func (w *Worker) startQueue(ctx context.Context, queue string, handler JobHandler) error {
	subject := fmt.Sprintf("worker.%s", queue)
	streamName := w.config.StreamName

	// Ensure stream exists
	if err := w.ensureStream(ctx, streamName, subject); err != nil {
		return fmt.Errorf("failed to ensure stream: %w", err)
	}

	// Create or update durable consumer
	durableName := fmt.Sprintf("blok-worker-%s", queue)
	_, err := w.js.CreateOrUpdateConsumer(ctx, streamName, jetstream.ConsumerConfig{
		Durable:       durableName,
		AckPolicy:     jetstream.AckExplicitPolicy,
		MaxDeliver:    w.config.MaxRetries + 1, // +1 for initial delivery
		AckWait:       w.config.AckWait + 5*time.Second, // Buffer for processing
		FilterSubject: subject,
	})
	if err != nil {
		return fmt.Errorf("failed to create consumer: %w", err)
	}

	// Get consumer handle
	consumer, err := w.js.Consumer(ctx, streamName, durableName)
	if err != nil {
		return fmt.Errorf("failed to get consumer: %w", err)
	}

	// Use a semaphore for concurrency control
	sem := make(chan struct{}, w.config.Concurrency)

	// Start consuming with callback
	cc, err := consumer.Consume(func(msg jetstream.Msg) {
		sem <- struct{}{} // Acquire semaphore

		go func() {
			defer func() { <-sem }() // Release semaphore

			w.processMessage(ctx, msg, queue, handler)
		}()
	})
	if err != nil {
		return fmt.Errorf("failed to start consumer: %w", err)
	}

	w.mu.Lock()
	w.subs = append(w.subs, cc)
	w.mu.Unlock()

	log.Printf("[Worker] Subscribed to queue: %s (stream=%s, consumer=%s)", queue, streamName, durableName)
	return nil
}

// processMessage handles a single message from NATS.
func (w *Worker) processMessage(ctx context.Context, msg jetstream.Msg, queue string, handler JobHandler) {
	// Extract headers
	headers := make(map[string]string)
	if msg.Headers() != nil {
		for key := range msg.Headers() {
			headers[key] = msg.Headers().Get(key)
		}
	}

	// Extract job metadata
	jobID := headers["x-job-id"]
	if jobID == "" {
		jobID = headers["Nats-Msg-Id"]
	}
	if jobID == "" {
		jobID = fmt.Sprintf("job-%d", time.Now().UnixNano())
	}

	// Get metadata for redelivery count
	meta, err := msg.Metadata()
	attempt := 0
	if err == nil && meta != nil {
		attempt = int(meta.NumDelivered) - 1 // 0-based
		if attempt < 0 {
			attempt = 0
		}
	}

	job := &JobMessage{
		ID:         jobID,
		Queue:      queue,
		Data:       msg.Data(),
		Headers:    headers,
		Attempt:    attempt,
		MaxRetries: w.config.MaxRetries,
	}

	log.Printf("[Worker] Processing job %s from %s (attempt %d/%d)", job.ID, queue, attempt+1, w.config.MaxRetries+1)

	start := time.Now()
	if err := handler(ctx, job); err != nil {
		elapsed := time.Since(start)
		log.Printf("[Worker] Job %s failed after %v: %v", job.ID, elapsed, err)

		// Nak for redelivery (NATS handles max_deliver limit)
		if nakErr := msg.Nak(); nakErr != nil {
			log.Printf("[Worker] Failed to nak message: %v", nakErr)
		}
		return
	}

	elapsed := time.Since(start)
	log.Printf("[Worker] Job %s completed in %v", job.ID, elapsed)

	// Acknowledge successful processing
	if ackErr := msg.Ack(); ackErr != nil {
		log.Printf("[Worker] Failed to ack message: %v", ackErr)
	}
}

// ensureStream creates or updates a JetStream stream.
func (w *Worker) ensureStream(ctx context.Context, name, subject string) error {
	// Try to get existing stream
	stream, err := w.js.Stream(ctx, name)
	if err == nil {
		// Stream exists — check if subject is included
		info, infoErr := stream.Info(ctx)
		if infoErr != nil {
			return infoErr
		}

		subjectExists := false
		for _, s := range info.Config.Subjects {
			if s == subject || s == subject+".*" || s == subject+".>" {
				subjectExists = true
				break
			}
		}

		if !subjectExists {
			// Add the new subject
			newSubjects := append(info.Config.Subjects, subject)
			cfg := info.Config
			cfg.Subjects = newSubjects
			_, updateErr := w.js.UpdateStream(ctx, cfg)
			if updateErr != nil {
				return fmt.Errorf("failed to update stream subjects: %w", updateErr)
			}
		}
		return nil
	}

	// Stream doesn't exist — create it
	_, createErr := w.js.CreateStream(ctx, jetstream.StreamConfig{
		Name:      name,
		Subjects:  []string{"worker.>"},
		Retention: jetstream.WorkQueuePolicy,
		Storage:   jetstream.FileStorage,
	})
	if createErr != nil {
		return fmt.Errorf("failed to create stream: %w", createErr)
	}

	log.Printf("[Worker] Created JetStream stream: %s", name)
	return nil
}

// Dispatch publishes a job to a worker queue.
// This is a convenience method for programmatic job dispatching.
func (w *Worker) Dispatch(ctx context.Context, queue string, data interface{}, opts *DispatchOpts) (string, error) {
	if w.nc == nil || w.js == nil {
		return "", fmt.Errorf("worker not connected, call Start() first")
	}

	subject := fmt.Sprintf("worker.%s", queue)

	// Encode data as JSON
	payload, err := json.Marshal(data)
	if err != nil {
		return "", fmt.Errorf("failed to marshal job data: %w", err)
	}

	// Build message with headers
	msg := nats.NewMsg(subject)
	msg.Data = payload

	jobID := fmt.Sprintf("job-%d", time.Now().UnixNano())
	if opts != nil && opts.JobID != "" {
		jobID = opts.JobID
	}

	msg.Header.Set("x-job-id", jobID)
	msg.Header.Set("Nats-Msg-Id", jobID) // Deduplication

	if opts != nil {
		if opts.Priority > 0 {
			msg.Header.Set("x-priority", strconv.Itoa(opts.Priority))
		}
		if opts.Delay > 0 {
			msg.Header.Set("x-delay", strconv.Itoa(int(opts.Delay.Milliseconds())))
		}
		if opts.Timeout > 0 {
			msg.Header.Set("x-timeout", strconv.Itoa(int(opts.Timeout.Milliseconds())))
		}
	}

	// Publish via JetStream
	_, pubErr := w.js.PublishMsg(ctx, msg)
	if pubErr != nil {
		return "", fmt.Errorf("failed to publish job: %w", pubErr)
	}

	return jobID, nil
}

// DispatchOpts holds optional job dispatch configuration.
type DispatchOpts struct {
	// JobID is an explicit job ID (for deduplication).
	JobID string

	// Priority is the job priority (higher = more important).
	Priority int

	// Delay is the delay before the job should be processed.
	Delay time.Duration

	// Timeout is the processing timeout for this job.
	Timeout time.Duration
}

// ListenAndServeWorker is a convenience function that creates a worker,
// auto-registers all nodes as queue handlers, and starts processing.
// It also starts an HTTP health server in the background.
func ListenAndServeWorker(registry *NodeRegistry) error {
	workerCfg := LoadWorkerConfigFromEnv()
	serverCfg := LoadConfigFromEnv()

	worker := NewWorker(registry, workerCfg)

	// Auto-register all nodes as queue handlers if queues are specified
	if len(workerCfg.Queues) > 0 {
		nodeNames := registry.NodeNames()
		for _, queue := range workerCfg.Queues {
			// If queue name matches a node name, auto-route to that node
			matched := false
			for _, name := range nodeNames {
				if queue == name {
					worker.HandleNode(queue, name)
					matched = true
					break
				}
			}
			// If no match, register a generic handler that routes based on job headers
			if !matched {
				worker.HandleNode(queue, queue)
			}
		}
	}

	// Start HTTP health server in background
	server := NewServer(registry, serverCfg)
	go func() {
		if err := server.Start(); err != nil {
			log.Printf("[Worker] Health server error: %v", err)
		}
	}()

	// Set up graceful shutdown
	ctx, cancel := context.WithCancel(context.Background())
	shutdownCh := SetupGracefulShutdown(func() {
		cancel()
		if err := server.Shutdown(); err != nil {
			log.Printf("[Worker] Error shutting down health server: %v", err)
		}
	})

	err := worker.Start(ctx)
	<-shutdownCh
	return err
}

File: sdks/rust/src/worker.rs
//! # NATS JetStream Worker
//!
//! Standalone background job processing for Blok workflows using NATS JetStream.
//!
//! ## Quick Start
//!
//! ```rust,no_run
//! use blok::worker::{Worker, WorkerConfig};
//! use blok::registry::NodeRegistry;
//!
//! #[tokio::main]
//! async fn main() -> Result<(), Box<dyn std::error::Error>> {
//!     let config = WorkerConfig::from_env();
//!     let registry = NodeRegistry::new(&config.version);
//!     let mut worker = Worker::new(registry, config);
//!     worker.handle_node("emails", "send-email");
//!     worker.start().await
//! }
//! ```

use std::collections::HashMap;
use std::env;
use std::sync::Arc;
use std::time::{Duration, Instant};

use async_nats::jetstream;
use async_nats::jetstream::consumer::PullConsumer;
use async_nats::jetstream::stream::RetentionPolicy;
use async_nats::jetstream::AckKind;
use tokio::sync::{Mutex, Semaphore};
use tracing::{error, info};

use crate::registry::NodeRegistry;
use crate::types::{Context, ExecutionRequest, NodeConfig, Request};

/// Worker configuration loaded from environment variables.
#[derive(Debug, Clone)]
pub struct WorkerConfig {
    /// NATS server URLs.
    pub servers: Vec<String>,

    /// JetStream stream name (default: "blok-worker").
    pub stream_name: String,

    /// Runtime version for health checks.
    pub version: String,

    /// Authentication token (optional).
    pub token: Option<String>,

    /// Username (optional).
    pub user: Option<String>,

    /// Password (optional).
    pub pass: Option<String>,

    /// Max concurrent job handlers (default: 1).
    pub concurrency: usize,

    /// Max delivery attempts (default: 3).
    pub max_retries: u32,

    /// Job timeout before NATS redelivers (default: 30s).
    pub ack_wait: Duration,

    /// Queue names to subscribe to.
    pub queues: Vec<String>,

    /// HTTP health server port (default: 8080).
    pub health_port: u16,
}

impl Default for WorkerConfig {
    fn default() -> Self {
        Self {
            servers: vec!["localhost:4222".into()],
            stream_name: "blok-worker".into(),
            version: "1.0.0".into(),
            token: None,
            user: None,
            pass: None,
            concurrency: 1,
            max_retries: 3,
            ack_wait: Duration::from_secs(30),
            queues: Vec::new(),
            health_port: 8080,
        }
    }
}

impl WorkerConfig {
    /// Load configuration from environment variables.
    ///
    /// - `NATS_SERVERS`: Comma-separated NATS server URLs (default: localhost:4222)
    /// - `NATS_TOKEN`: Authentication token
    /// - `NATS_USER`: Username
    /// - `NATS_PASS`: Password
    /// - `NATS_STREAM_NAME`: JetStream stream name (default: blok-worker)
    /// - `VERSION`: Runtime version (default: 1.0.0)
    /// - `WORKER_CONCURRENCY`: Max concurrent jobs (default: 1)
    /// - `WORKER_MAX_RETRIES`: Max delivery attempts (default: 3)
    /// - `WORKER_ACK_WAIT_SECS`: Job timeout in seconds (default: 30)
    /// - `WORKER_QUEUES`: Comma-separated queue names
    /// - `PORT`: HTTP health server port (default: 8080)
    pub fn from_env() -> Self {
        Self {
            servers: env::var("NATS_SERVERS")
                .map(|v| v.split(',').map(|s| s.trim().to_string()).collect())
                .unwrap_or_else(|_| vec!["localhost:4222".into()]),
            stream_name: env::var("NATS_STREAM_NAME").unwrap_or_else(|_| "blok-worker".into()),
            version: env::var("VERSION").unwrap_or_else(|_| "1.0.0".into()),
            token: env::var("NATS_TOKEN").ok(),
            user: env::var("NATS_USER").ok(),
            pass: env::var("NATS_PASS").ok(),
            concurrency: env::var("WORKER_CONCURRENCY")
                .ok()
                .and_then(|v| v.parse().ok())
                .unwrap_or(1),
            max_retries: env::var("WORKER_MAX_RETRIES")
                .ok()
                .and_then(|v| v.parse().ok())
                .unwrap_or(3),
            ack_wait: Duration::from_secs(
                env::var("WORKER_ACK_WAIT_SECS")
                    .ok()
                    .and_then(|v| v.parse().ok())
                    .unwrap_or(30),
            ),
            queues: env::var("WORKER_QUEUES")
                .map(|v| v.split(',').map(|s| s.trim().to_string()).collect())
                .unwrap_or_default(),
            health_port: env::var("PORT")
                .ok()
                .and_then(|v| v.parse().ok())
                .unwrap_or(8080),
        }
    }
}

/// A job received from a NATS worker queue.
#[derive(Debug, Clone)]
pub struct JobMessage {
    /// Unique job identifier.
    pub id: String,

    /// Queue name this job came from.
    pub queue: String,

    /// Raw JSON job payload.
    pub data: serde_json::Value,

    /// Message headers.
    pub headers: HashMap<String, String>,

    /// Current delivery attempt (0-based).
    pub attempt: u32,

    /// Maximum retries configured.
    pub max_retries: u32,
}

/// Handler type for processing jobs.
pub type JobHandlerFn =
    Arc<dyn Fn(JobMessage) -> std::pin::Pin<Box<dyn std::future::Future<Output = Result<(), Box<dyn std::error::Error + Send + Sync>>> + Send>> + Send + Sync>;

/// Worker processes background jobs from NATS JetStream queues.
pub struct Worker {
    config: WorkerConfig,
    registry: Arc<Mutex<NodeRegistry>>,
    handlers: HashMap<String, JobHandlerFn>,
}

impl Worker {
    /// Create a new NATS JetStream worker.
    pub fn new(registry: NodeRegistry, config: WorkerConfig) -> Self {
        Self {
            config,
            registry: Arc::new(Mutex::new(registry)),
            handlers: HashMap::new(),
        }
    }

    /// Register a custom async handler for a queue.
    pub fn handle<F, Fut>(&mut self, queue: &str, handler: F)
    where
        F: Fn(JobMessage) -> Fut + Send + Sync + 'static,
        Fut: std::future::Future<Output = Result<(), Box<dyn std::error::Error + Send + Sync>>>
            + Send
            + 'static,
    {
        let handler = Arc::new(move |job: JobMessage| {
            let fut = handler(job);
            Box::pin(fut)
                as std::pin::Pin<
                    Box<
                        dyn std::future::Future<
                                Output = Result<(), Box<dyn std::error::Error + Send + Sync>>,
                            > + Send,
                    >,
                >
        });
        self.handlers.insert(queue.to_string(), handler);
    }

    /// Register a handler that routes jobs to a registered node.
    pub fn handle_node(&mut self, queue: &str, node_name: &str) {
        let registry = self.registry.clone();
        let node_name = node_name.to_string();
        let max_retries = self.config.max_retries;

        self.handle(queue, move |job: JobMessage| {
            let registry = registry.clone();
            let node_name = node_name.clone();

            async move {
                let mut params = HashMap::new();
                params.insert("queue".to_string(), job.queue.clone());
                params.insert("jobId".to_string(), job.id.clone());
                params.insert("attempt".to_string(), job.attempt.to_string());

                let mut vars = HashMap::new();
                vars.insert(
                    "_worker_job".to_string(),
                    serde_json::json!({
                        "id": job.id,
                        "queue": job.queue,
                        "attempts": job.attempt.to_string(),
                        "maxRetries": max_retries.to_string(),
                    }),
                );

                let mut req = ExecutionRequest {
                    node: NodeConfig {
                        name: node_name,
                        path: String::new(),
                        node_type: String::new(),
                        config: match &job.data {
                            serde_json::Value::Object(m) => {
                                m.iter()
                                    .map(|(k, v)| (k.clone(), v.clone()))
                                    .collect()
                            }
                            _ => HashMap::new(),
                        },
                    },
                    context: Context {
                        id: job.id.clone(),
                        workflow_name: String::new(),
                        workflow_path: String::new(),
                        request: Request {
                            body: job.data,
                            headers: job.headers,
                            params,
                            query: HashMap::new(),
                            method: String::new(),
                            url: String::new(),
                            cookies: HashMap::new(),
                            base_url: String::new(),
                        },
                        response: Default::default(),
                        vars,
                        env: HashMap::new(),
                    },
                };

                let reg = registry.lock().await;
                let result = reg.execute(&mut req).await;

                if result.success {
                    Ok(())
                } else {
                    let msg = result
                        .errors
                        .as_ref()
                        .and_then(|e| e.get("message"))
                        .and_then(|v| v.as_str())
                        .unwrap_or("node execution failed");
                    Err(msg.to_string().into())
                }
            }
        });
    }

    /// Start the worker. Connects to NATS, ensures streams, and processes jobs.
    /// Blocks until a shutdown signal is received.
    pub async fn start(self) -> Result<(), Box<dyn std::error::Error>> {
        // Build NATS connection options
        let mut opts = async_nats::ConnectOptions::new();

        if let Some(token) = &self.config.token {
            opts = opts.token(token.clone());
        }
        if let (Some(user), Some(pass)) = (&self.config.user, &self.config.pass) {
            opts = opts.user_and_password(user.clone(), pass.clone());
        }

        // Connect to NATS
        let servers = self.config.servers.join(",");
        let client = opts.connect(&servers).await.map_err(|e| {
            format!("Failed to connect to NATS at {}: {}", servers, e)
        })?;

        info!("[Worker] Connected to NATS: {}", servers);

        // Create JetStream context
        let js = jetstream::new(client.clone());

        // Ensure stream exists
        let stream = ensure_stream(&js, &self.config).await?;

        // Start processing each queue
        let mut tasks = Vec::new();
        let sem = Arc::new(Semaphore::new(self.config.concurrency));

        for (queue, handler) in &self.handlers {
            let subject = format!("worker.{}", queue);
            let durable_name = format!("blok-worker-{}", queue);

            // Create or update consumer
            let consumer: PullConsumer = stream
                .create_consumer(jetstream::consumer::pull::Config {
                    durable_name: Some(durable_name.clone()),
                    ack_policy: jetstream::consumer::AckPolicy::Explicit,
                    max_deliver: self.config.max_retries as i64 + 1,
                    ack_wait: self.config.ack_wait + Duration::from_secs(5),
                    filter_subject: subject.clone(),
                    ..Default::default()
                })
                .await
                .map_err(|e| format!("Failed to create consumer for {}: {}", queue, e))?;

            info!(
                "[Worker] Subscribed to queue: {} (stream={}, consumer={})",
                queue, self.config.stream_name, durable_name
            );

            // Spawn a task that continuously pulls messages
            let handler = handler.clone();
            let sem = sem.clone();
            let queue = queue.clone();
            let max_retries = self.config.max_retries;

            let task = tokio::spawn(async move {
                let mut messages = consumer
                    .messages()
                    .await
                    .expect("Failed to get message stream");

                use tokio_stream::StreamExt;

                while let Some(msg_result) = messages.next().await {
                    let msg = match msg_result {
                        Ok(m) => m,
                        Err(e) => {
                            error!("[Worker] Error receiving message from {}: {}", queue, e);
                            continue;
                        }
                    };

                    // Acquire semaphore permit for concurrency control
                    let permit = sem.clone().acquire_owned().await;
                    if permit.is_err() {
                        break; // Semaphore closed
                    }
                    let _permit = permit.unwrap();

                    // Extract headers
                    let mut headers = HashMap::new();
                    if let Some(hdrs) = msg.headers.as_ref() {
                        for (key, values) in hdrs.iter() {
                            if let Some(val) = values.iter().next() {
                                headers.insert(key.to_string(), val.to_string());
                            }
                        }
                    }

                    // Extract job ID
                    let job_id = headers
                        .get("x-job-id")
                        .or_else(|| headers.get("Nats-Msg-Id"))
                        .cloned()
                        .unwrap_or_else(|| format!("job-{}", chrono::Utc::now().timestamp_millis()));

                    // Get delivery count from message info
                    let info = msg.info();
                    let attempt = info
                        .as_ref()
                        .map(|i| i.delivered.saturating_sub(1) as u32)
                        .unwrap_or(0);

                    // Parse message data
                    let data: serde_json::Value =
                        serde_json::from_slice(&msg.payload).unwrap_or(serde_json::Value::Null);

                    let job = JobMessage {
                        id: job_id.clone(),
                        queue: queue.clone(),
                        data,
                        headers,
                        attempt,
                        max_retries,
                    };

                    info!(
                        "[Worker] Processing job {} from {} (attempt {}/{})",
                        job_id,
                        queue,
                        attempt + 1,
                        max_retries + 1
                    );

                    let start = Instant::now();
                    let handler = handler.clone();

                    // Execute handler
                    match handler(job).await {
                        Ok(()) => {
                            let elapsed = start.elapsed();
                            info!("[Worker] Job {} completed in {:?}", job_id, elapsed);

                            if let Err(e) = msg.ack().await {
                                error!("[Worker] Failed to ack message: {}", e);
                            }
                        }
                        Err(e) => {
                            let elapsed = start.elapsed();
                            error!(
                                "[Worker] Job {} failed after {:?}: {}",
                                job_id, elapsed, e
                            );

                            // Nak for redelivery (NATS handles max_deliver limit)
                            if let Err(nak_err) = msg.ack_with(AckKind::Nak(None)).await {
                                error!("[Worker] Failed to nak message: {}", nak_err);
                            }
                        }
                    }
                }
            });

            tasks.push(task);
        }

        info!(
            "[Worker] Processing {} queue(s), concurrency={}",
            tasks.len(),
            self.config.concurrency
        );

        // Wait for shutdown signal
        tokio::signal::ctrl_c()
            .await
            .expect("Failed to listen for ctrl-c");

        info!("[Worker] Shutting down...");

        // Abort all tasks
        for task in tasks {
            task.abort();
        }

        // Drain NATS connection
        client.flush().await?;

        info!("[Worker] Stopped");
        Ok(())
    }
}

/// Ensure the JetStream stream exists.
async fn ensure_stream(
    js: &jetstream::Context,
    config: &WorkerConfig,
) -> Result<jetstream::stream::Stream, Box<dyn std::error::Error>> {
    // Try to get existing stream
    match js.get_stream(&config.stream_name).await {
        Ok(stream) => {
            info!(
                "[Worker] Using existing JetStream stream: {}",
                config.stream_name
            );
            Ok(stream)
        }
        Err(_) => {
            // Create stream with WorkQueue retention
            let stream = js
                .create_stream(jetstream::stream::Config {
                    name: config.stream_name.clone(),
                    subjects: vec!["worker.>".into()],
                    retention: RetentionPolicy::WorkQueue,
                    storage: jetstream::stream::StorageType::File,
                    ..Default::default()
                })
                .await
                .map_err(|e| {
                    format!(
                        "Failed to create JetStream stream '{}': {}",
                        config.stream_name, e
                    )
                })?;

            info!(
                "[Worker] Created JetStream stream: {}",
                config.stream_name
            );
            Ok(stream)
        }
    }
}
